src/rocksdb_options.rs

```diff
@@ -1246,11 +1245,32 @@ impl ColumnFamilyOptions {
                 Ok(s) => s,
                 Err(e) => return Err(format!("failed to convert to cstring: {:?}", e)),
             };
-            self.filter = Some(new_compaction_filter(c_name, ignore_snapshots, filter)?);
-            crocksdb_ffi::crocksdb_options_set_compaction_filter(
-                self.inner,
-                self.filter.as_ref().unwrap().inner,
-            );
+            let filter = new_compaction_filter(c_name, filter);
+            crocksdb_ffi::crocksdb_options_set_compaction_filter(self.inner, filter.inner);
+            self.filter = Some(filter);
+            Ok(())
+        }
+    }
+
+    /// Set compaction filter factory.
+    ///
+    /// See also `CompactionFilter`.
+    pub fn set_compaction_filter_factory<S>(
+        &mut self,
+        name: S,
+        factory: Box<dyn CompactionFilterFactory>,
+    ) -> Result<(), String>
+    where
+        S: Into<Vec<u8>>,
+    {
+        let c_name = match CString::new(name) {
+            Ok(s) => s,
+            Err(e) => return Err(format!("failed to convert to cstring: {:?}", e)),
+        };
+        unsafe {
+            let factory = new_compaction_filter_factory(c_name, factory)?;
+            crocksdb_ffi::crocksdb_options_set_compaction_filter_factory(self.inner, factory.inner);
+            std::mem::forget(factory); // Avoid call destructor.
```

Comment:
Can we hold the factory like filter?

Reply:
But we do not need the factory. Maybe we could merge this two function together? `new_compaction_filter_factory ` and `crocksdb_options_set_compaction_filter_factory `.

Reply:
I prefer to remove `filter` from `ColumnFamilyOptions`.

Reply:
`filter` can't be removed from `ColumnFamilyOptions` because in `rocksdb::ColumnFamilyOptions` it's a raw pointer. We need to manage its lifetime by ourself.
However `compaction_filter_factory` in `rocksdb::ColumnFamilyOptions` is a shared pointer, so it's ok.